### PR TITLE
fix: allow tsc to typecheck tests, fix type issues in those tests

### PR DESCRIPTION
**Pull request by etrepum**
Mostly just adding non-null assertions and adding types where inference fails. I did refactor a few list reduces that do a concat (which is also quadratic time) to use flatMap which is easier for TypeScript to infer.

I added some docs to the FAQ relevant to reconciliation and how the tests are written https://lexical-git-fork-etrepum-fix-test-types-fbopensource.vercel.app/docs/faq#when-does-reconciliation-happen

### CODE REVIEW

File: packages/lexical-history/src/__tests__/unit/LexicalHistory.test.tsx [hunk 1]
```diff
@@ -34,12 +34,12 @@ import {
 } from 'lexical/src';
 import {createTestEditor, TestComposer} from 'lexical/src/__tests__/utils';
 import React from 'react';
-import {createRoot} from 'react-dom/client';
+import {createRoot, Root} from 'react-dom/client';
```
Comment (potatowagon) [inline packages/lexical-history/src/__tests__/unit/LexicalHistory.test.tsx:37]: nit: we could import Root as type instead
```suggestion
import type {Root} from 'react-dom/client';
import {createRoot} from 'react-dom/client';
```

File: packages/lexical-react/src/__tests__/unit/Collaboration.test.ts [hunk 2]
```diff
@@ -8,22 +8,22 @@
 
 import {$createTextNode, $getRoot, ParagraphNode, TextNode} from 'lexical';
 
-import {createTestConnection, waitForReact} from './utils';
+import {Client, createTestConnection, waitForReact} from './utils';
```
Comment (potatowagon) [inline packages/lexical-react/src/__tests__/unit/Collaboration.test.ts:11]:
```suggestion
import type {Client} from './utils';
import {createTestConnection, waitForReact} from './utils';
```

File: packages/lexical-react/src/__tests__/unit/LexicalComposer.test.tsx [hunk 3]
```diff
@@ -8,14 +8,14 @@
 
 import {useLexicalComposerContext} from '@lexical/react/LexicalComposerContext';
 import * as React from 'react';
-import {createRoot} from 'react-dom/client';
+import {createRoot, Root} from 'react-dom/client';
```
Comment (potatowagon) [inline packages/lexical-react/src/__tests__/unit/LexicalComposer.test.tsx:11]:
```suggestion
import type {Root} from 'react-dom/client';
import {createRoot} from 'react-dom/client';
```

File: packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx [hunk 4]
```diff
@@ -22,9 +22,10 @@ import {
   $getRoot,
   $getSelection,
   $isRangeSelection,
+  LexicalEditor,
 } from 'lexical';
 import * as React from 'react';
-import {createRoot} from 'react-dom/client';
+import {createRoot, Root} from 'react-dom/client';
```
Comment (potatowagon) [inline packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx:28]:
```suggestion
import type {Root} from 'react-dom/client';
import {createRoot} from 'react-dom/client';
```

File: packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx [hunk 5]
```diff
@@ -22,9 +22,10 @@ import {
   $getRoot,
   $getSelection,
   $isRangeSelection,
+  LexicalEditor,
 } from 'lexical';
```
Comment (potatowagon) [inline packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx:26]:
```suggestion
} from 'lexical';
import type {LexicalEditor} from 'lexical';
```

File: packages/lexical-selection/src/__tests__/unit/LexicalSelection.test.tsx [hunk 6]
```diff
@@ -30,7 +30,13 @@ import {
   $isElementNode,
   $isRangeSelection,
   $setSelection,
+  DecoratorNode,
+  ElementNode,
+  LexicalEditor,
+  LexicalNode,
```
Comment (potatowagon) [inline packages/lexical-selection/src/__tests__/unit/LexicalSelection.test.tsx:36]: ditto nit: could be imported as type instead

File: packages/lexical-website/docs/faq.md [hunk 7]
```diff
@@ -37,6 +37,55 @@ If you've used React Hooks before, you can think of `$` functions as being somet
...
+```
+
+At a high level, very roughly, the order of operations looks like this:
+
+1. `editor.update()` is called
+2. `updateA()` is called and updates the editor state
+3. `editor.update()` schedules a reconciliation microtask and returns
+4. `await` schedules a resume microtask and yields control to the task executor
+5. the reconciliation microtask runs, reconciling the editor state with the DOM
+6. the resume microtask runs
```
Comment (potatowagon) [inline packages/lexical-website/docs/faq.md:87]: 👍 
thanks for adding this! will leave to @zurfyx and other reconciler experts to vet

File: packages/lexical-selection/src/__tests__/utils/index.ts [hunk 8]
```diff
@@ -12,6 +12,8 @@ import {
   $isNodeSelection,
   $isRangeSelection,
   $isTextNode,
+  LexicalEditor,
+  PointType,
 } from 'lexical';
```
Comment (potatowagon) [inline packages/lexical-selection/src/__tests__/utils/index.ts:17]:
```suggestion
} from 'lexical';

import type {LexicalEditor, PointType} from 'lexical';
```

**Review by potatowagon — COMMENTED** [review]
- the only non test file modified is `packages/lexical-selection/src/__tests__/utils/index.ts`. adds typing only so low risk change
- rest of code changes to test files and adds typing -> very low risk change
- left some minor nits about importing as type

looks good! thankyou for this :)

**Review by potatowagon — DISMISSED** [review]

**Comment by etrepum** [comment]
I'd be happy to fix up those imports if you want, I was just doing whatever the language server suggested since I was fixing a few hundred type errors! Perhaps it would be better to do that cleanup separately in a more systematic way? There are other bad import patterns that we could get rid of in the tests (e.g. importing public code from 'lexical/src' instead of 'lexical', and we could get rid of those tsconfig hacks too). It seems likely that we are also importing values as types elsewhere in there… if that's even something we care about? Type imports are usually done as a performance thing or to prevent import side-effects or circular dependencies if values from the module are not otherwise imported.

**Comment by etrepum** [comment]
If we do clean up the type-only imports there are linters and typescript flags that should be configured to enforce whatever style is chosen, e.g. https://typescript-eslint.io/rules/consistent-type-imports/  and https://www.typescriptlang.org/tsconfig/#verbatimModuleSyntax are relevant for that

**Comment by potatowagon** [comment]
yes addressing the type imports in a seperate PR makes sense

would be great if a lint could autofix this

**Review by StyleT — APPROVED** [review]

**Comment by etrepum** [comment]
Latest commit just fixes the conflicts with main

**Review by potatowagon — APPROVED** [review]

**Review by StyleT — APPROVED** [review]

**Pull request merged**